### DISCUSSION TITLE
Request to update documentation regarding Custom Providers

### DISCUSSION THREAD

**Discussion by denik1981** [nextauthjs/next-auth#3237] [Help]
I found that there is a big mismatch in the Typed Provider the object is expecting and the one shown in the documentation. 
Also is it possible to append a tiny description about each of the fields requested for a custom provider? It will help us know how to proceded when building a new one.

**Comment by balazsorban44** [comment]
Curious about the big mismatch.

Are you thinking of this: https://next-auth.js.org/configuration/providers/oauth#options

**Reply by denik1981** [reply]
From [Using a Custom Provider](https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider) docs:
)
```ts
{
  id: "google",
  name: "Google",
  type: "oauth",
  wellKnown: "https://accounts.google.com/.well-known/openid-configuration",
  authorization: { params: { scope: "openid email profile" } },
  idToken: true,
  checks: ["pkce", "state"],
  profile(profile) {
    return {
      id: profile.sub,
      name: profile.name,
      email: profile.email,
      image: profile.picture,
    }
  },
}
```


From [providers.d.ts](https://github.com/nextauthjs/next-auth/blob/main/types/providers.d.ts)
```ts
export interface OAuthConfig<P extends Record<string, unknown> = Profile>
  extends CommonProviderOptions {
  authorizationParams?: Record<string, string>
  headers?: Record<string, any>
  type: "oauth"
  version: string
  scope: string | string[]
  params: { grant_type: string }
  accessTokenUrl: string
  requestTokenUrl?: string
  authorizationUrl: string
  profileUrl: string
  profile(profile: P, tokens: TokenSet): Awaitable<User & { id: string }>
  protection?: ProtectionType | ProtectionType[]
  clientId: string
  clientSecret:
    | string
    // TODO: only allow for Apple
    | Record<"appleId" | "teamId" | "privateKey" | "keyId", string>
  idToken?: boolean
  /**
   * @deprecated Will be removed in an upcoming major release. Use `protection: ["state"]` instead.
   */
  state?: boolean

  // TODO: only allow for BattleNet
  region?: string
  // TODO: only allow for some
  domain?: string
  // TODO: only allow for Azure Active Directory B2C and FusionAuth
  tenantId?: string
}
```

**Reply by denik1981** [reply]
The referenced example needs a revision cause that object isn't asignable to the type next-auth expects for a new custom provider.
Options docs are ok.

**Reply by balazsorban44** [reply]
you link to the wrong branch. that is for v3, the docs is currently the v4 version.

See this section of the migration guide explaining the changes https://next-auth.js.org/getting-started/upgrade-v4#providers

**Reply by denik1981** [reply]
Ok, I can see that. Thanks for pointing that out but please check that I was linking to v4 as well (not v3).

[https://next-auth.js.org/getting-started/upgrade-v4#providers](https://next-auth.js.org/getting-started/upgrade-v4#providers)
[https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider](https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider)

Both of the above links below to v4 docs. Maybe the second one could be updated or include a warning sign for checking the breaking changes for v4  mentioned in the upgrade section. I was following the example code on v4 docs to build my custom provider and was a little bit confussed about why the type interface was so different.
